Disable checkpointing properly if openPMD is missing

ikbuibui opened the pull request:
- [x]  #5492 needs to be merged before this.

Moves checkpointing simulation control in its own class
Checkpoint class now has a compile time toggle to disable checkpoint/restart functionality (fixes #5480)
Simplified restart state management and changed from using bools to an enum tracking state

ikbuibui committed:
Make pmacc checkpointing class

Moves checkpointing simulation control in its own class
Checkpoint class now has a compile time toggle to disable
checkpoint/restart functionality
Simplified restart state management and changed from using bools to an
enum tracking state

ikbuibui commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,374 @@
+                }
+
+                // Checkpoints are expected to be sorted chronologically.
+                bool const stepFound = std::binary_search(checkpoints.cbegin(), checkpoints.cend(), restartStep);
```
I have this logic here

chillenzer commented on include/picongpu/simulation/control/Simulation.hpp:
```diff
@@ -69,6 +70,7 @@
 #include <pmacc/particles/traits/FilterByIdentifier.hpp>
 #include <pmacc/random/RNGProvider.hpp>
 #include <pmacc/random/methods/methods.hpp>
+#include <pmacc/simulationControl/Checkpointing.hpp>
```
I would have expected `Checkpoint.hpp` but I presume you are distinguishing it from a single checkpoint (like referencing iteration 4892)?

ikbuibui replied:
Yeah, this was the main cause of my unease.

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,440 @@
+        {
+            // Handle unknown enum values gracefully.
+            out << "UNKNOWN_RESTART_STATE(" << static_cast<int>(state) << ")";
+        }
```
Let's hope, we've never gotta do this!

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,440 @@
+               {"try", RestartState::TRY},
+               {"force", RestartState::FORCE},
+               {"success", RestartState::SUCCESS},
+               {"failed", RestartState::FAILED}};
```
Should we extract this? Having to update this in multiple places seems bound to fail. On the other hand, we'll hopefully not really have to update this ever.

ikbuibui replied:
Unified

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,440 @@
+        else
+        {
+            // If the token is not a valid state, set the stream's failbit.
+            // boost::program_options will catch this and report an error to the user.
```
Wow! Dark magic. Why not just throw yourself here?

ikbuibui replied:
I prefer to leave the responsibility of IO to boost po

chillenzer replied:
Is this an idiomatic and documented way to use boost po? If so, could you please point me to a reference? Haven't found it anywhere in [the docs](https://www.boost.org/doc/libs/latest/doc/html/program_options.html).

ikbuibui replied:
Using a custom validator is most likely the idiomatic way to do it (discussed in the SO question below), but I chose the (IMO) simpler option.
I took over the solution from [here](https://stackoverflow.com/questions/27526817/input-stream-operator-lookup-order-with-boost-program-options)

ikbuibui replied:
I can add a validate as well if you think it helps

chillenzer replied:
If that's not an officially supported protocol, I'd refrain from using this obscure way to signal a failure. If it's not official interface, boost po would be allowed to change this behaviour without further notice, e.g., just swallow the failbit and keep going. I can see the advantages of the "initialisation = validation" approach that Pydantic is taking as well but I wouldn't do so at the expense of relying on implementation details of a code we don't control. 

My suggestion (admittedly I haven't checked the `validate` protocol, so please fill in the blanks):
```C++
std::optional<RestartState> stringToRestartState(std::string const given) {
    // ...
    return it == ...end()? std::nullopt : it->first;
}

inline auto validate(std::istream& in, RestartState& state) {
    // ...
    auto state = stringToRestartState(token);
    if (not state.has_value()) {
        // ...
    }
    return ...
}

inline std::istream& operator>>(std::istream& in, RestartState& state) {
    auto state = stringToRestartState(token);
    // ...
}
```
(Also I'm presuming that this is not at all performance critical or anything, so not taking care of how many copies we make here. etc.)

ikbuibui replied:
I was able to use `validate` and get away without defining the stream operators at all as there is an overload of `implicit_value` which takes the string representation.

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,440 @@
+                return true;
+            }
+            return false;
+        }
```
Not guarded by `checkpointingEnabled`.

ikbuibui replied:
We allowed restart attempts, even without checkpoints (from the start of the simulation). This only maintains old functionality.

ikbuibui replied:
I moved it to this class just to keep the user facing `checkpoint.restart.loop` input

chillenzer replied:
Okay. Not 100% happy with the code duplication that happened due to separating the template specialisations but I'll let that slip.

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,440 @@
+     */
+    template<bool checkpointingEnabled>
+    struct Checkpointing
+    {
```
I strongly dislike the approach of guarding each individual function body with an `if constexpr` because you have to repeat this over and over again. I would have expected to just provide a specialisation
```C++
template<>
struct Checkpointing<false> {
    void registerHelp(...) {}
    ...
};
```
Is there a strong reason for your design?

ikbuibui replied:
Used your suggestion

chillenzer replied:
Looks good. Two tiny things:
- Shall we declare the default class template without a definition and define only the specialisations?
```C++
template<bool checkpointingEnabled>
struct Checkpointing;

template<>
structu Checkpointing<true> {
...
```
- Boolean interfaces are an anti-pattern (see e.g. [this explanation and example](https://docs.astral.sh/ruff/rules/boolean-type-hint-positional-argument/)). I'd prefer to have this as an enum.
```C++
enum class CheckpointingAvailability {
...

template<CheckpointingAvailability checkpointingEnabled>
struct Checkpointing {
...
```

Just to be explicit: The second without the first would not be a good idea because the second opens the possibility of more modes.

chillenzer replied:
PS: Not blocking.

ikbuibui replied:
Changed

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,440 @@
+                // clang-format on
+                // translate checkpointPeriod string into checkpoint intervals
+                seqCheckpointPeriod = pluginSystem::toTimeSlice(checkpointPeriod);
+            }
```
Diff'ed against the removed parts from `SimulationHelper` and looks correct.

chillenzer commented on include/pmacc/simulationControl/SimulationHelper.cpp:
```diff
@@ -27,52 +27,33 @@
-        if(checkpointTimeThread.joinable())
-            checkpointTimeThread.join();
-
+        checkpointing.endTimeBasedCheckpointing();
```
Wow, camel case is not great here. I had to dig into the code to realise that it's not an apocalytic "end-time-based checkpointing" but it's just "ending the time-based checkpointing". Could we rename to "finish"/"cancel"/...?

(I would have been curious what an apocalyptic "end-time-based checkpointing" could look like, though, if you ever want to implement one...)

ikbuibui replied:
renamed to `finishTimeBasedCheckpointing`

chillenzer reviewed (changes requested):
Partial review.

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,440 @@
+                return true;
+            }
+            return false;
+        }
```
Please rename this function. Starting with `do...` suggests heavily that it actually `do`es something. In the `while` context it's used in it doesn't read too bad but out of context it's weird. You might want to try something along the lines of `moreSoftRestartAttemptsAllowed`.

ikbuibui replied:
renamed to `canSoftRestart`. But as I type it now i think this name has the `empty` problem. It is not clear if the caller asking or telling

ikbuibui replied:
now changed to the more verbose  `tryConsumeRestartAttempt`

chillenzer replied:
I don't like that you still have a verb in the name that could be read as an imperative because it still conveys the meaning of doing something. Options would be any of the following:
- third person forms of a verb, particularly `is...`, `has...` but also `allows...`
- any participle `...allowed`, `...left`, etc.
- any attribute `...available`

Suggestions:
```C++
while(checkpointing.softRestartAttemptsAreLeft())
while(checkpointing.hasSoftRestartAttemptsLeft())
while(checkpointing.allowsSoftRestartAttempts())
```

ikbuibui replied:
using `hasSoftRestartAttemptsLeft`

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,440 @@
+                            }
+                        });
+            }
+        }
```
Same here: `do...` suggests that it does something while it actually rather `start...`s or `activate...`s.

ikbuibui replied:
renamed to `startTimeBasedCheckpointing` to complement the `finishTimeBasedCheckpointing`

chillenzer reviewed (changes requested):
Looks okay to me otherwise. Please do the marked renamings.

PrometheusPi commented:
@ikbuibui what is the status of this PR?

ikbuibui commented:
I still need to address the requested changes, will happen sometime this week.

ikbuibui committed:
Apply review comments

Specialized the Checkpointing class template for disabled, i.e. false
type
Did some renamings
Unified the enum to string mapping

ikbuibui committed:
review round 2

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,508 @@
+            }
+
+            // Checkpoints are expected to be sorted chronologically.
+            bool const stepFound = std::binary_search(checkpoints.cbegin(), checkpoints.cend(), restartStep);
```
This assertion is wrong, I believe: You've got the `addCheckpoint` which will push a random number to the end of the vector. Even if that's not an immediate bug currently due to the precise path through the code that's taken to arrive here, I still think that the performance gain at this point is not worth risking this to become a problem.

ikbuibui replied:
This is a good point, I now replaced this with a more resilient reversed find, since I think the most recent matching checkpoint should be preferred over older ones (in the event such a case occurs)

chillenzer commented on include/pmacc/simulationControl/Checkpointing.hpp:
```diff
@@ -0,0 +1,508 @@
+        }
+
+        /* Returns whether a restart needs to be performed */
+        bool checkRestart(uint32_t& step)
```
If you have that comment right above, why not?
```suggestion
        bool restartNeedsToBePerformed(uint32_t& step)
```

ikbuibui replied:
`restartNeedsToBePerformed` also sounds like an instruction to do something

chillenzer replied:
I see... I was thinking about
```C++
if(checkpointing.restartNeedsToBePerformed)
```
which reads quite clearly but that might be just to my ears. This is not blocking. We can keep it as is.

chillenzer reviewed (approved):
Thanks for bearing with me for so long. We finally made it!
